File: Backend/app/provider-platform/static-offer-driver-app/Main/src/Tools/Notifications.hs
 {-
 Copyright 2022-23, Juspay India Pvt Ltd
 
 This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License 
 
 as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version. This program 
 
 is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
 
 or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details. You should have received a copy of 
 
 the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
-}

module Tools.Notifications where

import qualified Data.Text as T
import Domain.Types.Booking (Booking)
import qualified Domain.Types.BookingCancellationReason as SBCR
import Domain.Types.Person as Person
import Domain.Types.RegistrationToken as RegToken
import EulerHS.Prelude
import qualified Kernel.External.FCM.Flow as FCM
import Kernel.External.FCM.Types as FCM
import Kernel.Storage.Hedis (HedisFlow)
import Kernel.Tools.Metrics.CoreMetrics (CoreMetrics)
import Kernel.Types.Error
import Kernel.Types.Id
import Kernel.Utils.Common

-- | Send FCM "cancel" notification to driver
notifyOnCancel ::
  ( HedisFlow m r,
    CoreMetrics m,
    MonadFlow m
  ) =>
  FCMConfig ->
  Booking ->
  Id Person ->
  Maybe FCM.FCMRecipientToken ->
  SBCR.CancellationSource ->
  m ()
notifyOnCancel fcmConfig booking personId mbDeviceToken cancellationSource = do
  cancellationText <- getCancellationText
  FCM.notifyPerson fcmConfig (notificationData cancellationText) $ FCMNotificationRecipient personId.getId mbDeviceToken
  where
    notificationData cancellationText =
      FCM.FCMData
        { fcmNotificationType = FCM.CANCELLED_PRODUCT,
          fcmShowNotification = FCM.SHOW,
          fcmEntityType = FCM.Product,
          fcmEntityIds = getId booking.id,
          fcmEntityData = (),
          fcmNotificationJSON = FCM.createAndroidNotification title (body cancellationText) FCM.CANCELLED_PRODUCT
        }
    title = FCMNotificationTitle $ T.pack "Ride cancelled!"
    body text =
      FCMNotificationBody text
    getCancellationText = case cancellationSource of
      SBCR.ByUser ->
        return $
          unwords
            [ "Customer had to cancel your ride for",
              showTimeIst (booking.startTime) <> ".",
              "Check the app for more details."
            ]
      SBCR.ByMerchant ->
        return $
          unwords
            [ "Your agency had to cancel the ride for",
              showTimeIst (booking.startTime) <> ".",
              "Check the app for more details."
            ]
      SBCR.ByDriver ->
        return $
          unwords
            [ "You have cancelled the ride for",
              showTimeIst (booking.startTime) <> ".",
              "Check the app for more details."
            ]
      _ -> throwError (InternalError "Unexpected cancellation reason.")

notifyOnRegistration ::
  ( HedisFlow m r,
    CoreMetrics m,
    MonadFlow m
  ) =>
  FCMConfig ->
  RegistrationToken ->
  Id Person ->
  Maybe FCM.FCMRecipientToken ->
  m ()
notifyOnRegistration fcmConfig regToken personId =
  FCM.notifyPerson fcmConfig notificationData . FCMNotificationRecipient personId.getId
  where
    tokenId = RegToken.id regToken
    notificationData =
      FCM.FCMData
        { fcmNotificationType = FCM.REGISTRATION_APPROVED,
          fcmShowNotification = FCM.SHOW,
          fcmEntityType = FCM.Merchant,
          fcmEntityIds = getId tokenId,
          fcmEntityData = (),
          fcmNotificationJSON = FCM.createAndroidNotification title body FCM.REGISTRATION_APPROVED
        }
    title = FCMNotificationTitle $ T.pack "Registration Completed!"
    body =
      FCMNotificationBody $
        unwords
          [ "Welcome Yatri Partner!",
            "Click here to set up your account."
          ]

notifyDriver ::
  ( HedisFlow m r,
    CoreMetrics m,
    MonadFlow m
  ) =>
  FCMConfig ->
  FCM.FCMNotificationType ->
  Text ->
  Text ->
  Id Person ->
  Maybe FCM.FCMRecipientToken ->
  m ()
notifyDriver fcmConfig = sendNotificationToDriver fcmConfig FCM.SHOW Nothing

-- Send notification to device, i.e. notifications that should not be shown to the user,
-- but contains payload used by the app
notifyDevice ::
  ( HedisFlow m r,
    CoreMetrics m,
    MonadFlow m
  ) =>
  FCMConfig ->
  FCM.FCMNotificationType ->
  Text ->
  Text ->
  Id Person ->
  Maybe FCM.FCMRecipientToken ->
  m ()
notifyDevice fcmConfig = sendNotificationToDriver fcmConfig FCM.DO_NOT_SHOW (Just FCM.HIGH)

sendNotificationToDriver ::
  ( HedisFlow m r,
    CoreMetrics m,
    MonadFlow m
  ) =>
  FCMConfig ->
  FCM.FCMShowNotification ->
  Maybe FCM.FCMAndroidMessagePriority ->
  FCM.FCMNotificationType ->
  Text ->
  Text ->
  Id Person ->
  Maybe FCM.FCMRecipientToken ->
  m ()
sendNotificationToDriver fcmConfig displayOption priority notificationType notificationTitle message driverId =
  FCM.notifyPersonWithPriority fcmConfig priority notificationData . FCMNotificationRecipient driverId.getId
  where
    notificationData =
      FCM.FCMData
        { fcmNotificationType = notificationType,
          fcmShowNotification = displayOption,
          fcmEntityIds = getId driverId,
          fcmEntityType = FCM.Person,
          fcmEntityData = (),
          fcmNotificationJSON = FCM.createAndroidNotification title body notificationType
        }
    title = FCM.FCMNotificationTitle notificationTitle
    body =
      FCMNotificationBody message

notifyDriverNewAllocation ::
  ( HedisFlow m r,
    CoreMetrics m,
    MonadFlow m
  ) =>
  FCMConfig ->
  Id bookingId ->
  Id Person ->
  Maybe FCM.FCMRecipientToken ->
  m ()
notifyDriverNewAllocation fcmConfig bookingId personId =
  FCM.notifyPersonWithPriority fcmConfig (Just FCM.HIGH) notificationData . FCMNotificationRecipient personId.getId
  where
    title = FCM.FCMNotificationTitle "New allocation request."
    body =
      FCM.FCMNotificationBody $
        unwords
          [ "New ride request!",
            "Check the app for more details."
          ]
    notificationData =
      FCM.FCMData
        { fcmNotificationType = FCM.ALLOCATION_REQUEST,
          fcmShowNotification = FCM.SHOW,
          fcmEntityType = FCM.Product,
          fcmEntityIds = getId bookingId,
          fcmEntityData = (),
          fcmNotificationJSON = FCM.createAndroidNotification title body FCM.ALLOCATION_REQUEST
        }

notifyRideNotAssigned ::
  ( HedisFlow m r,
    CoreMetrics m,
    MonadFlow m
  ) =>
  FCMConfig ->
  Id Booking ->
  Id Person ->
  Maybe FCM.FCMRecipientToken ->
  m ()
notifyRideNotAssigned fcmConfig bookingId personId =
  FCM.notifyPerson fcmConfig notificationData . FCMNotificationRecipient personId.getId
  where
    title = FCM.FCMNotificationTitle "Ride not assigned."
    body =
      FCM.FCMNotificationBody $
        unwords
          [ "Could not assign the ride as it is no longer available",
            "Please wait for another request."
          ]
    notificationData =
      FCM.FCMData
        { fcmNotificationType = FCM.ALLOCATION_REQUEST_UNASSIGNED,
          fcmShowNotification = FCM.SHOW,
          fcmEntityType = FCM.Product,
          fcmEntityIds = getId bookingId,
          fcmEntityData = (),
          fcmNotificationJSON = FCM.createAndroidNotification title body FCM.ALLOCATION_REQUEST_UNASSIGNED
        }

notifyFarePolicyChange ::
  ( HedisFlow m r,
    CoreMetrics m,
    MonadFlow m
  ) =>
  FCMConfig ->
  Id coordinatorId ->
  Maybe FCM.FCMRecipientToken ->
  m ()
notifyFarePolicyChange fcmConfig coordinatorId =
  FCM.notifyPerson fcmConfig notificationData . FCMNotificationRecipient coordinatorId.getId
  where
    title = FCM.FCMNotificationTitle "Fare policy changed."
    body =
      FCM.FCMNotificationBody $
        unwords
          [ "Fare has been updated."
          ]
    notificationData =
      FCM.FCMData
        { fcmNotificationType = FCM.FARE_POLICY_CHANGED,
          fcmShowNotification = FCM.SHOW,
          fcmEntityType = FCM.Person,
          fcmEntityIds = getId coordinatorId,
          fcmEntityData = (),
          fcmNotificationJSON = FCM.createAndroidNotification title body FCM.FARE_POLICY_CHANGED
        }

notifyDiscountChange ::
  ( HedisFlow m r,
    CoreMetrics m,
    MonadFlow m
  ) =>
  FCMConfig ->
  Id coordinatorId ->
  Maybe FCM.FCMRecipientToken ->
  m ()
notifyDiscountChange fcmConfig coordinatorId =
  FCM.notifyPerson fcmConfig notificationData . FCMNotificationRecipient coordinatorId.getId
  where
    title = FCM.FCMNotificationTitle "Discount updated."
    body =
      FCM.FCMNotificationBody $
        unwords
          [ "Discount has been changed."
          ]
    notificationData =
      FCM.FCMData
        { fcmNotificationType = FCM.DISCOUNT_CHANGED,
          fcmShowNotification = FCM.SHOW,
          fcmEntityType = FCM.Person,
          fcmEntityIds = getId coordinatorId,
          fcmEntityData = (),
          fcmNotificationJSON = FCM.createAndroidNotification title body FCM.DISCOUNT_CHANGED
        }

File: Frontend/ios-native/Namma Yatri/Notifications.swift
/* 
 *  Copyright 2022-23, Juspay India Pvt Ltd
 *  This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License
 *  as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version. This program
 *  is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 *  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details. You should have received a copy of
 *  the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

//
//  Notifications.swift
//  Namma Yatri
//
//  Created by Balaganesh on 03/02/23.
//

import UIKit

extension AppDelegate {
    
    // For notification with alert & sound
    func userNotificationCenter(_ center: UNUserNotificationCenter,
                                  willPresent notification: UNNotification,
                                  withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions)
                                    -> Void) {
        let userInfo = notification.request.content.userInfo
        #if DEBUG
        print("userInfo: \(userInfo)")
        #endif
        let info = self.extractUserInfoFrom(dict: userInfo)
        let notificationType = info?.notificationType as? String ?? ""
        let createdAt = info?.createdAt as? String ?? ""
        if (notificationType == "REGISTRATION_APPROVED") {
            storeInHyperSDKUserDefaultsWith(key: notificationType, value: "true")
        }
        if let viewController = self.window?.rootViewController as? ViewController {
            viewController.hyperInstance.callCallback(forNotificationType: notificationType);
        }
        if (notificationType == "DRIVER_QUOTE_INCOMING" && !shouldAllowNotification(createdAt : createdAt)) {
            return;
        }
        completionHandler([[.alert, .sound]])
    }
    
    // For silent push notification
    func application(_ application: UIApplication,
                     didReceiveRemoteNotification userInfo: [AnyHashable : Any],
                     fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
        
        let info = self.extractUserInfoFrom(dict: userInfo)
        #if DEBUG
        print("userInfo: \(userInfo)")
        #endif
        if let viewController = self.window?.rootViewController as? ViewController {
            let notificationType = info?.notificationType as? String ?? ""
            viewController.hyperInstance.callCallback(forNotificationType: notificationType);
        }
        
        // Inform the system after the background operation is completed.
        completionHandler(.newData)
    }
    
    
    func extractUserInfoFrom(dict: [AnyHashable : Any]) -> UserInfo? {
        guard let aps = dict["aps"] as? [String: Any] else {return nil}
        guard let alert = aps["alert"] as? [String: Any] else {return nil}
        guard let data = aps["data"] as? [String: Any] else {return nil}
        guard let entityDataString = data["entity_data"] as? String else {return nil}
        guard let entityData = entityDataString.data(using: .utf8) else {return nil}
        guard let entityDataJSON = try? JSONSerialization.jsonObject(with: entityData) as? Array<Any?> else {return nil}
         
        let title = alert["title"] as? String ?? ""
        let body = alert["body"] as? String ?? ""
        let notificationType = aps["category"] as? String ?? ""
        var createdAt = ""
        if let firstEntity = entityDataJSON.first as? [String: Any] {
            createdAt = firstEntity["createdAt"] as? String ?? ""
        }
         
        let info = UserInfo(title : title, body : body, notificationType : notificationType, createdAt : createdAt)
        return info
    }
    
    // For checking the specific NotificationType: "DRIVER_QUOTE_INCOMING"
    // Returns true if notification can be allowed
    func shouldAllowNotification(createdAt: String) -> Bool{
        let createdAtDate = createdAt.replacingOccurrences(of: "T", with: " ").replacingOccurrences(of: "Z", with: "")
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd HH:mm:ss.SSSSSSSSS"
        let currentDate = dateFormatter.date(from: createdAtDate)
        
        guard let previousDate = self.previousDate else {
            self.previousDate = currentDate
            return true
        }
        
        guard let currentDate = currentDate else {
            return true
        }
        
        let diffDate = currentDate.timeIntervalSinceReferenceDate - previousDate.timeIntervalSinceReferenceDate
        if (diffDate > 30) {
            self.previousDate = currentDate
            return true
        }
        return false
    }
}
